src/Security/VoterV2/PersonalDataVoter.php

```diff
@@ -38,8 +40,32 @@ protected function voteOnAttribute(string $attribute, $subject, TokenInterface $
             return false;
         }
 
-        return $user->isBeneficiaire()
-            ? $this->checker->isGranted('UPDATE', $subject->getBeneficiaire())
-            : $this->checker->isGranted('UPDATE', $subject->getBeneficiaire()) && false === $subject->getBPrive();
+        return match ($attribute) {
+            self::UPDATE => $this->canUpdate($user, $subject),
+            self::TOGGLE_VISIBILITY => $this->canToggleVisibility($user, $subject),
+            default => false,
+        };
+    }
+
+    private function canUpdate(User $user, DonneePersonnelle $subject): bool
+    {
+        if (!$this->checker->isGranted('UPDATE', $subject->getBeneficiaire())) {
+            return false;
+        }
+
+        if ($user->isMembre() && $subject->getBPrive()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private function canToggleVisibility(User $user, DonneePersonnelle $subject): bool
+    {
+        if ($subject instanceof FolderableEntityInterface) {
+            return !$subject->hasParentFolder() && $this->canUpdate($user, $subject);
+        }
+
+        return $this->canUpdate($user, $subject);
```

Comment:
this feels strange, maybee you could merge statements
```php
if (!this->canUpdate()) {
    return false;
}

if ($subject instanceof FolderableEntityInterface) {
    return !$subject->hasParentFolder()
 }

return true;
```

You can even shortcut the last lines by switching statements

```php
if ($subject instanceof FolderableEntityInterface && !$subject->hasParentFolder() ) {
    return false;
 }

return this->canUpdate()
```